README.md
# About
[![License BSD 3-Clause](https://img.shields.io/badge/license-BSD%203--Clause-blue.svg)](http://lesscss-extended-compiler.projects.gabrys.biz/license.txt)
[![Build Status](https://travis-ci.org/gabrysbiz/lesscss-extended-compiler.svg?branch=master)](https://travis-ci.org/gabrysbiz/lesscss-extended-compiler)

Extended version of the [LessCSS Compiler](http://lesscss-compiler.projects.gabrys.biz/)
(Java library which compiles [Less](http://lesscss.org/) source files to the [CSS](http://www.w3.org/Style/CSS/) code).
Adds features such as:
* caching ([see caching data](http://lesscss-extended-compiler.projects.gabrys.biz/2.1.0/caching-data.html)):
 * compiled code
 * source code
* supports sources located at:
 * local drives
 * network protocols: HTTP, HTTPS, FTP
 * classpath (prefix `classpath://`)
 * [custom - defined by programmers](http://lesscss-extended-compiler.projects.gabrys.biz/2.1.0/locations.html#custom)
 
# Compatibility
The compatibility with the [Less](http://lesscss.org/) language is determined by:
* native compiler implementation ([biz.gabrys.lesscss.compiler.LessCompiler](http://lesscss-compiler.projects.gabrys.biz/1.2.1/apidocs/index.html?biz/gabrys/lesscss/compiler/LessCompiler.html))
* supported locations of sources ([see supported locations matrix](http://lesscss-extended-compiler.projects.gabrys.biz/2.1.0/locations.html#matrix))

# Requirements
The compiler to run requires:
* Java 5.0 or higher
* Third-Party Dependencies ([see list](http://lesscss-extended-compiler.projects.gabrys.biz/2.1.0/dependencies.html))

# Download
You can download the library from [this page](http://lesscss-extended-compiler.projects.gabrys.biz/2.1.0/download.html)
or using various [dependency management tools](http://lesscss-extended-compiler.projects.gabrys.biz/2.1.0/dependency-info.html).


CHANGELOG.md
# Extended LessCSS Compiler Changelog

## 2.1.0
New Features:
* Classpath "protocol" support

API:
* [AbstractSourceTreePreparationProcessor#expirationChecker](http://lesscss-extended-compiler.projects.gabrys.biz/2.1.0/apidocs/index.html?biz/gabrys/lesscss/extended/compiler/control/processor/AbstractSourceTreePreparationProcessor.html) is no longer `final`
* [AbstractSourceTreePreparationProcessor#importsCache](http://lesscss-extended-compiler.projects.gabrys.biz/2.1.0/apidocs/index.html?biz/gabrys/lesscss/extended/compiler/control/processor/AbstractSourceTreePreparationProcessor.html) is no longer `final`
* [AbstractSourceTreePreparationProcessor#sourceFactory](http://lesscss-extended-compiler.projects.gabrys.biz/2.1.0/apidocs/index.html?biz/gabrys/lesscss/extended/compiler/control/processor/AbstractSourceTreePreparationProcessor.html) is no longer `final`
* [FullCacheImpl#storage](http://lesscss-extended-compiler.projects.gabrys.biz/2.1.0/apidocs/index.html?biz/gabrys/lesscss/extended/compiler/cache/FullCacheImpl.html) is no longer `final`

[See documentation](http://lesscss-extended-compiler.projects.gabrys.biz/2.1.0/)

## 2.0
New Features:
* [FTP](https://tools.ietf.org/html/rfc959) protocol support

API:
* [DataStorage#put(fileName, lines)](http://lesscss-extended-compiler.projects.gabrys.biz/2.0/apidocs/index.html?biz/gabrys/lesscss/extended/compiler/storage/DataStorage.html): changed the type of the parameter `lines` from `java.lang.Iterable` to `java.util.Collection`
* [ConcreteSourceFactory](http://lesscss-extended-compiler.projects.gabrys.biz/2.0/apidocs/index.html?biz/gabrys/lesscss/extended/compiler/source/ConcreteSourceFactory.html) is now generic: `ConcreteSourceFactory<T extends LessSource>`

Dependencies:
* Use <code>commons-lang3</code> instead of <code>commons-lang</code>

Bugs:
* Security fixes:
  * changed the way of creating temporary directories (see [OWAPS Top Ten 2013 Category A9](https://www.owasp.org/index.php/Top_10_2013-A9-Using_Components_with_Known_Vulnerabilities))
  * protection against NULL bytes (see [CWE-158](http://cwe.mitre.org/data/definitions/158.html), [WASC-28](http://projects.webappsec.org/w/page/13246949/Null%20Byte%20Injection))
* Optimization of create new collections

[See documentation](http://lesscss-extended-compiler.projects.gabrys.biz/2.0/)

## 1.0
Initial release.

[See documentation](http://lesscss-extended-compiler.projects.gabrys.biz/1.0/)

src/main/java/biz/gabrys/lesscss/extended/compiler/source/LocalSource.java
/*
 * Extended LessCSS Compiler
 * http://lesscss-extended-compiler.projects.gabrys.biz/
 *
 * Copyright (c) 2015 Adam Gabryś
 *
 * This file is licensed under the BSD 3-Clause (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain:
 *  - a copy of the License at project page
 *  - a template of the License at https://opensource.org/licenses/BSD-3-Clause
 */
package biz.gabrys.lesscss.extended.compiler.source;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.Date;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.SystemUtils;

/**
 * Represents a <a href="http://lesscss.org/">Less</a> source stored in local file system.
 * @since 1.0
 */
public class LocalSource implements LessSource {

    private final File file;
    private final String encoding;

    private Date lastModificationDate;

    /**
     * Constructs a new instance and sets file which store source code with default platform encoding.
     * @param file the file which store source code.
     * @since 1.0
     */
    public LocalSource(final File file) {
        this(file, Charset.defaultCharset().toString());
    }

    /**
     * Constructs a new instance and sets file which store source code and its encoding.
     * @param file the file which store source code.
     * @param encoding the source file encoding.
     * @since 1.0
     */
    public LocalSource(final File file, final String encoding) {
        this.file = file;
        this.encoding = encoding;
    }

    public String getPath() {
        try {
            return file.getCanonicalPath();
        } catch (final IOException e) {
            throw new SourceException(String.format("Cannot resolve canonical path for \"%s\" file", file.getAbsolutePath()), e);
        }
    }

    public String getEncoding() {
        return encoding;
    }

    public String getContent() {
        String content;
        try {
            content = FileUtils.readFileToString(file, encoding);
        } catch (final IOException e) {
            throw new SourceException(String.format("Cannot read content of the \"%s\" file", file.getAbsolutePath()), e);
        }
        lastModificationDate = new Date(file.lastModified());
        return content;
    }

    /**
     * {@inheritDoc} Before the first {@link #getContent()} call returns current source modification time. After the
     * first {@link #getContent()} call always returns the source modification time read while reading the source
     * contents.
     * @since 1.0
     */
    public Date getLastModificationDate() {
        if (lastModificationDate != null) {
            return (Date) lastModificationDate.clone();
        }
        return new Date(file.lastModified());
    }

    /**
     * Checks whether the path represents an absolute path. Always returns {@code false} if path contains NULL bytes
     * (see <a href="http://cwe.mitre.org/data/definitions/158.html">CWE-158</a>,
     * <a href="http://projects.webappsec.org/w/page/13246949/Null%20Byte%20Injection">WASC-28</a>).
     * @param path the checked path.
     * @return {@code true} whether the path represents an absolute path, otherwise {@code false}.
     * @since 1.0
     */
    public static boolean isAbsolutePath(final String path) {
        return isAbsolutePath(path, SystemUtils.IS_OS_WINDOWS);
    }

    static boolean isAbsolutePath(final String path, final boolean windowsOperatingSystem) {
        if (path != null && path.indexOf('\0') > -1) {
            return false;
        }

        final String normalizedPath = FilenameUtils.normalize(path, true);
        if (StringUtils.isEmpty(normalizedPath)) {
            return false;
        }
        return isAbsoluteNormalizedPath(normalizedPath, windowsOperatingSystem);
    }

    static boolean isAbsoluteNormalizedPath(final String normalizedPath, final boolean windowsOperatingSystem) {
        final char separator = '/';
        final char firstChar = normalizedPath.charAt(0);
        if (!windowsOperatingSystem) {
            return firstChar == separator;
        }

        if (!Character.isLetter(firstChar)) {
            return false;
        }

        final int colonIndex = normalizedPath.indexOf(':');
        final int separatorIndex = 2;
        return colonIndex == 1 && normalizedPath.length() > separatorIndex && normalizedPath.charAt(separatorIndex) == separator;
    }
}
